tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -219,49 +195,255 @@ static StatusOr<absl::optional<se::blas::AlgorithmType>> DoUncachedGemmAutotune(
+                                    output_buffer);
+
+    static const int64_t max_scratch_size = tensorflow::GetBlasWorkspaceLimit(
+        "TF_CUBLAS_WORKSPACE_LIMIT_IN_MB", 1LL << 32);  // 4GB by default
+    DCHECK(output_matrix.transpose == se::blas::Transpose::kNoTranspose);
+    tensorflow::DataType dtype;
+    se::blas::DataType blas_dtype;
+    switch (output_shape.element_type()) {
+      case F16:
+        dtype = tensorflow::DataTypeToEnum<Eigen::half>::value;
+        blas_dtype = se::blas::ToDataType<Eigen::half>::value;
+        break;
+      case F32:
+        dtype = tensorflow::DataTypeToEnum<float>::value;
+        blas_dtype = se::blas::ToDataType<float>::value;
+        break;
+      case F64:
+        dtype = tensorflow::DataTypeToEnum<double>::value;
+        blas_dtype = se::blas::ToDataType<double>::value;
+        break;
+      case C64:
+        dtype = tensorflow::DataTypeToEnum<complex64>::value;
+        blas_dtype = se::blas::ToDataType<complex64>::value;
+        break;
+      case C128:
+        dtype = tensorflow::DataTypeToEnum<complex128>::value;
+        blas_dtype = se::blas::ToDataType<complex128>::value;
+        break;
+    }
 
-  if (it != autotune_cache.end()) {
-    cache_hits++;
-    VLOG(4) << "Autotuning cache hit, using algorithm: "
-            << (it->second.has_value() ? absl::StrCat(*(it->second))
-                                       : "<generic>");
-    return it->second;
-  }
-  cache_misses++;
-  VLOG(4) << "Autotuning cache miss";
+    bool allow_tf32 = tensorflow::tensor_float_32_execution_enabled();
+    int device_id = stream->parent()->device_ordinal();
+    bool trans_x = lhs_matrix.transpose == se::blas::Transpose::kTranspose;
+    bool trans_y = rhs_matrix.transpose == se::blas::Transpose::kTranspose;
+
+    int64_t m = output_matrix.num_rows;
+    int64_t n = output_matrix.num_cols;
+    auto k = lhs_matrix.reduced_dim();
+
+    bool broadcast = batch_size == 1;
+    tensorflow::BatchMatmulParameters matmul_parameters(
+        trans_x, trans_y, false, false, m, n, k, batch_size,
+        /*broadcast_a*/ broadcast, /*broadcast_b*/ broadcast, dtype, dtype,
+        allow_tf32, device_id);
+    static const int64_t max_autotune_algorithm_count =
+        tensorflow::MatmulMaxAutotuneAlgorithmCount();
+
+    const auto* plan_and_algorithms =
+        tensorflow::BatchMatmulPlanMapSingleton::GetInstance()->Find(
+            matmul_parameters);
+    if (!plan_and_algorithms) {
+      se::blas::BlasLtMatmulPlanParams plan_params =
+          CreatePlanParams(batch_size, blas_dtype, dtype, lhs_matrix,
+                           rhs_matrix, output_matrix)
+              .ConsumeValueOrDie();
+
+      auto status_or_plan =
+          stream->parent()->CreateBlasLtMatmulPlan(plan_params);
+      TF_RETURN_IF_ERROR(tensorflow::FromExecutorStatus(status_or_plan));
+      std::unique_ptr<se::blas::IBlasLtMatmulPlan> plan =
```

Comment:
At this point it seems we have too many levels of indentation, a function would help